src/main/java/ai/preferred/venom/Crawler.java

```diff
@@ -271,14 +325,44 @@ private void run() {
             LOGGER.debug("The thread pool is interrupted");
             return;
           }
-          fetcher.fetch(crawlerRequest, new AsyncCrawlerCallbackProcessor(this, job));
+
+          final CompletableFuture<Response> completableResponseFuture = new CompletableFuture<>();
+          completableResponseFuture
+              .whenComplete((response, throwable) -> connections.release())
+              .thenAcceptAsync(response -> handle(job, response), threadPool)
+              .whenComplete((response, throwable) -> {
+                if (throwable != null) {
+                  final Throwable cause = throwable.getCause();
```

Comment:
Can we forget about the throwable itself here?

Reply:
Exceptions are propagated from the top.

If an exception is thrown prior to line 332, i.e. from fetcher, thenAcceptAsync will not run, and the exception from the fetcher is handled on line 333.

If an exception is thrown on line 332, the exception can also be handled on line 333. But we catch all on line 332, so there should be nothing thrown.